lambda/transl_list_comprehension.ml

```diff
@@ -0,0 +1,302 @@
+      let element = Ident.create_local "element" in
+      { builder      = rev_dlist_concat_map
+      ; arg_lets     = [iter_list]
+      ; element
+      ; element_kind = Typeopt.value_kind pattern.pat_env pattern.pat_type
+      ; add_bindings =
+          (* CR aspectorzabusky: This has to be at [value_kind] [Pgenval],
+             right, since we don't know more specifically? *)
+          Matching.for_let
+            ~scopes pattern.pat_loc (Lvar element) pattern Pgenval
+      }
+
+(** Translates a list comprehension binding
+    ([Typedtree.comprehension_clause_binding]) into Lambda.  At parse time,
+    iterators don't include patterns and bindings do; however, in the typedtree
+    representation, the patterns have been moved into the iterators (so that
+    range iterators can just have an [Ident.t], for translation into for loops),
+    so bindings are just like iterators with a possible annotation.  As a
+    result, this function is essentially the same as [iterator], which see. *)
+let binding ~transl_exp ~scopes { comp_cb_iterator; comp_cb_attributes = _ } =
+  (* CR aspectorzabusky: What do we do with attributes here? *)
+  iterator ~transl_exp ~scopes comp_cb_iterator
+
+(** Translate all the bindings of a single [for ... and ...] clause (the
+    contents of a [Typedtree.Texp_comp_for]) into a pair of (1) a list of let
+    bindings that are in force for the translation; and (2) a single Lambda term
+    of type ['res rev_dlist], assuming we know how to translate everything that
+    ought to be nested within it (the [inner_body], a function awaiting the most
+    nested accumulator as a labeled argument which will produce the body of the
+    iterations) and have a name for the accumulator of the current [rev_dlist]
+    ([accumulator], which changes at every recursive step).  It folds together
+    all the [translated_iterator]s by connecting their [body_func]tions to each
+    other, and bottoms out at the [inner_body].  *)
+let rec translate_bindings
+          ~transl_exp ~scopes ~loc ~inner_body ~accumulator = function
+  | cur_binding :: bindings ->
+      let { builder; arg_lets; element; element_kind; add_bindings } =
+        binding ~transl_exp ~scopes cur_binding
+      in
+      let inner_acc = Ident.create_local "accumulator" in
+      let body_arg_lets, body =
+        translate_bindings
+          ~transl_exp ~scopes ~loc
+          ~inner_body ~accumulator:(Lvar inner_acc) bindings
+      in
+      let body_func =
+        (* CR aspectorzabusky: How do I make this function stack allocate its result? *)
+        Lfunction
+          { kind   = Curried { nlocal = 1 } (* The accumulator is local *)
+          ; params = [element, element_kind; inner_acc, Pgenval]
+          ; return = Pgenval
+          ; attr   = default_function_attribute
+          ; loc
+          ; mode   = alloc_local
+          ; region = true
+              (* CR aspectorzabusky: We want one region per iterator, right? *)
```

Comment:
I don't see why we would...

Reply:
I think I don't have a good intuition here.  Say more?

Reply:
Well, because regions are invisible to users, it's important to let them get some intuition behind them. In other words, copy whatever is done for `for` loops, as we should just do the same here. That might be one region per iterator, indeed.

Reply:
Verdict: Do whatever `for` loops do

Reply:
I think I've done this, but generally I think it would be good if somebody who understood modes (@lpw25 or @stedolan) took a look at all the mode stuff to make sure it's coherent.